### PR TITLE
Add hostname for top site, dev logger, deduping

### LINKED ISSUE #180: Use the first letter of host for icon fallback
My embedly data isn't always showing up so this appears to be the very last fallback, but I think these should be the first letter of the host (B and G) rather than the first letter of the title (1 and 5).

screen shot 2016-02-23 at 8 22 30 am

### LINKED ISSUE #182: Duplicate sites in the same section
pasted_image_2_23_16__10_13_am

my top sites show twin links to invision.

one for www.invision.com and the other for invision.com.

### LINKED ISSUE #194: Invalid default favicon color in TopSites.js
Per https://github.com/mozilla/activity-streams/pull/126#discussion_r53349286 and [./content-src/components/TopSites/TopSites.js:20](https://github.com/mozilla/activity-streams/blob/95390eb65232c0e64087faa841174fd8def58575/content-src/components/TopSites/TopSites.js#L20), we're setting a fallback color of `rgb(333, 333, 333)`:

```js
const color = site.favicon_colors && 
        site.favicon_colors[0] &&
        site.favicon_colors[0].color ||
        [333, 333, 333];
```

Not sure that's a valid color, or if it should be something a bit less "whiter than the sun's core" white.

/cc @bryanbell for design feedback/approval

### PR DESCRIPTION
**Pull request by k88hudson**
This uses `url-parse` to dedupe sites (i.e. remove sites that are basically equivalent). We also need to do some work to dedupe *across* components, but this is a good start

Fixes #180 
Fixes #182 
Fixes #194

### CODE REVIEW

**Comment by k88hudson** [comment]
@rlr R?

**Comment by pdehaan** [comment]
```
SUMMARY:
✔ 99 tests completed
✖ 2 tests failed

FAILED TESTS:
  urlFilter
    ✖ should pass valid urls
      Firefox 45.0.0 (Linux 0.0.0)
    expected [] to deeply equal [ Array(3) ]
    ... 
   
    ✖ should remove localhost urls, but not urls that start with localhost
      Firefox 45.0.0 (Linux 0.0.0)
    expected [] to deeply equal [ { url: '/home/travis/build/mozilla/activity-streams/content-test/index.js:357:19 <- webpack:///~/assertion-error/index.js:73:0
    ...
```

File: content-src/lib/utils.js [hunk 1]
```diff
@@ -7,6 +7,12 @@ module.exports = {
     const yiq = ((r * 299) + (g * 587) + (b * 114)) / 1000;
     return (yiq >= 128) ? "black" : "white";
   },
+  prettyHostname(hostname) {
+    if (!hostname) {
+      return "";
+    }
+    return hostname.toLowerCase().replace(/^www./i, "");
+  },
```
Comment (pdehaan) [inline content-src/lib/utils.js]: This looks similar to the `prettyUrl()` regex below. Not sure if it makes sense to try and write a single regex/function which removes an optional "http[s]:", "//" and "www.".

```js
function prettyUrl(url) {
  if (!url) {
    return "";
  }
  return url.replace(/^((https?:)?\/\/)?(www\.)?/i, "").toLowerCase();
}
```

Then again, RegExes are an ugly thing, so maybe duplicate+simple is better than complex+ugly.

---

```
function prettyUrl(url) {
  if (!url) {
    return "";
  }
  return url.replace(/^((https?:)?\/\/)?(www\.)?/i, "").toLowerCase();
}

var sites = [
  "http://peter.com",
  "https://www.peter.com",
  "http://zombo.Com",
  "www.peter.com",
  "www.devnull.biz",
  "bleepbloop.com",
  "ftp://bananas.com",
  "//whatever.man"
];

sites.forEach((site) => {
  console.log("%s ==> %s", site, prettyUrl(site));
});

**Comment by k88hudson** [comment]
@pdehaan yeah, that's a good point

**Comment by pdehaan** [comment]
So far this looks awesome! Here's some random browsing data:

new_tab

File: content-src/components/TopSites/TopSites.js [hunk 2]
```diff
@@ -17,14 +17,28 @@ const TopSites = React.createClass({
...
-          const color = site.favicon_colors && site.favicon_colors[0] && site.favicon_colors[0].color || [333, 333, 333];
+          let title;
+          let color;
+          try {
+            title = prettyHostname(site.parsedUrl.hostname);
+          } catch (e) {
+            //
+          }
+          if (!title) {
+            title = site.provider_name || site.title;
```
Comment (pdehaan) [inline content-src/components/TopSites/TopSites.js]: Why not just move this bit up into the `let` assignment?

```js
let title = site.provider_name || site.title;

try {
  title = prettyHostname(site.parsedUrl.hostname);
} catch (e) {
 // ignore
}
```

Same with `color`, I suppose.
```js
let color = [251, 251, 251];

try {
  color = site.favicon_colors[0].color;
} catch (e) {
  // ignore
}
```

**Comment by rlr** [comment]
I like @pdehaan suggestions. Looks good otherwise. Travis not happy with some tests.

File: content-src/lib/dedupe.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+const {log} =  require("lib/logger");
+const {prettyHostname} = require("lib/utils");
+
+function createDedupeKey(site) {
+  const parsed = site.parsedUrl;
+  if (!parsed) {
+    return null;
+  }
+  const host = prettyHostname(parsed.host);
```
Comment (rlr) [inline content-src/lib/dedupe.js]: do we need to wrap this on in a `try{} catch{}` ?

File: content-src/components/TopSites/TopSites.js (hunk 2 above)
Comment (k88hudson) [inline content-src/components/TopSites/TopSites.js]: much better :+1:

**Pull request merged**